src/protocol/command.rs

```diff
@@ -1,10 +1,95 @@
 use super::argument_parser::Parser;
 use super::executable_path;
 use crate::R;
+use regex::Regex;
 use std::ffi::OsString;
 use std::path::PathBuf;
 use std::str;
 
+#[derive(Debug, Clone)]
+pub enum CommandMatcher {
+    Exact(Command),
+    RegexMatch(Regex),
+}
+
+impl PartialEq for CommandMatcher {
+    fn eq(&self, other: &CommandMatcher) -> bool {
+        use CommandMatcher::*;
+        match (self, other) {
+            (Exact(a), Exact(b)) => a == b,
+            (RegexMatch(a), RegexMatch(b)) => a.as_str() == b.as_str(),
+            _ => false,
+        }
+    }
+}
+
+impl Eq for CommandMatcher {}
+
+impl CommandMatcher {
+    pub fn matches_received(&self, received: &Command) -> bool {
+        match self {
+            CommandMatcher::Exact(command) => {
+                executable_path::compare_executables(&command.executable, &received.executable)
+                    && command.arguments == received.arguments
+            }
+            CommandMatcher::RegexMatch(regex) => regex.is_match(&received.format()),
+        }
+    }
+
+    pub fn format(&self) -> String {
+        match self {
+            CommandMatcher::Exact(command) => command.format(),
+            CommandMatcher::RegexMatch(regex) => regex.as_str().to_string(),
+        }
+    }
+
+    pub fn regex_match(string: &str) -> R<CommandMatcher> {
+        Ok(CommandMatcher::RegexMatch(Regex::new(&format!(
+            "^{}$",
```

Comment:
unfortunately the `Regex` create doesn't offer control over anchoring via the `RegexBuilder`, so i couldn't find any other way to get this in there. FWIW, even if the protocol includes anchors, the `Regex` ignores consecutive ones (ie, `^^\d$$` seems to be the same as `^\d$`)

Reply:
Can we have a test (or two) for the claim that it ignores consecutive anchors?